internal/sftp/goclient.go

```diff
@@ -57,6 +57,32 @@ func (c *GoClient) Upload(ctx context.Context, src io.Reader, dest string) (int6
 	return bytes, nil
 }
 
+// Delete removes the data from dest and returns the
+// number of bytes removed. A new SFTP connection is opened before removing the file, and
+// closed when the delete is complete or cancelled.
+//
+// Delete is not thread safe.
+func (c *GoClient) Delete(ctx context.Context, dest string) (int64, error) {
+	if err := c.dial(); err != nil {
+		return 0, err
+	}
+	defer c.close()
+
+	// Get file information to find out its size
+	fileInfo, err := c.sftp.Lstat(dest)
```

Comment:
I think even without returning `bytes` the lstat is a good check that the file exists before deleting it, so we can provide a better error message.

Reply:
An alternative where we do it in go to save us the extra round-trip:

```go
func (c *GoClient) Delete(ctx context.Context, dest string) error {
	if err := c.dial(); err != nil {
		return fmt.Errorf("SFTP: unable to dial: %v", err)
	}
	defer c.close()

	err := c.sftp.Remove(dest)
	if err != nil {
		// Wrap only non-exist and permission errors to avoid leaking.
		if os.IsNotExist(err) || os.IsPermission(err) {
			return fmt.Errorf("SFTP: unable to remove file %q: %w", dest, err)
		}
		return fmt.Errorf("SFTP: unable to remove file %q: %v", dest, err)
	}

	return nil
}
```

Notice that I'm only using `%w` when I want to expose `os.ErrNotExist` or `os.ErrPermission`, but using `%v` to avoid other errors that could lead to implementation leaking.

Reply:
I like this alternative, but I would make a small tweak to the non-exist/permission error message:
```
return fmt.Errorf("SFTP: unable to remove file: %w", err)
```
I find that most (maybe all?) of the filesystem error messages include the name of the file, so adding it in the wrapper causes duplication.  That said, it's easy to test for `os.IsNotExist(err)` in the unit tests so it's probably worth checking what the actual error message looks like.

Reply:
You're right. I had a quick look at the code and it looks like it only returns a sentinel error but it's better to test it.